carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/utils/FeatureFileGeneratorUtils.java

```diff
@@ -0,0 +1,452 @@
+            }
+        }
+        //endregion
+
+        try {
+            TransformerFactory transformerFactory = TransformerFactory.newInstance();
+            Transformer transformer;
+            transformer = transformerFactory.newTransformer();
+            DOMSource source = new DOMSource(document);
+            StreamResult result = new StreamResult(featureXmlFile);
+            transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+            transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
+            transformer.transform(source, result);
+        } catch (TransformerConfigurationException e) {
+            throw new TransformerConfigurationException("Unable to create feature manifest", e);
+        }
+    }
+
+    /**
+     * If a manifest file is given, parse the manifest file and return the Document object representing the file.
+     * Generates a new Document otherwise.
+     *
+     * @param manifest java.io.File pointing an existing manifest file.
+     * @return Document object representing a given manifest file or a newly generated manifest file
+     * @throws ParserConfigurationException
+     * @throws SAXException
+     * @throws IOException
+     */
+    private static Document getManifestDocument(File manifest) throws ParserConfigurationException,
+            SAXException, IOException {
+        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder documentBuilder;
+        try {
+            documentBuilder = documentBuilderFactory.newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            throw new ParserConfigurationException("Unable to load feature manifest");
+        }
+        Document document;
+        if (manifest != null && manifest.exists()) {
+            try (InputStream manifestFileStream = new FileInputStream(manifest)) {
+                document = documentBuilder.parse(manifestFileStream);
+            } catch (SAXException e) {
+                throw new SAXException("Unable to parse feature manifest", e);
+            } catch (IOException e) {
+                throw new IOException("Unable to load feature manifest", e);
+            }
+        } else {
+            document = documentBuilder.newDocument();
+        }
+        return document;
+    }
+
+    /**
+     * Cross check plugins given in the manifest file against the plugins configured in the pom.xml file. Returns a
+     * list of bundles found in the pom.xml but not in the manifest file.
+     *
+     * @param bundles  list of bundles configured in the pom.xml
+     * @param document Document representing the give manifest
+     * @return ArrayList&lt;Bundle&gt; missing plugins
+     */
+    private static ArrayList<Bundle> getMissingPlugins(List<Bundle> bundles, Document document) {
+        if (bundles == null || bundles.size() == 0) {
+            return new ArrayList<>();
```

Comment:
What about making the return statement use  "Collections.emptyList()" instead of instantiating a new ArrayList? I hope the returned list is not modified anywhere else in the code.